src/Server/HTTPHandler.cpp

```diff
@@ -818,7 +818,13 @@ void HTTPHandler::processQuery(
 
     /// While still no data has been sent, we will report about query execution progress by sending HTTP headers.
     /// Note that we add it unconditionally so the progress is available for `X-ClickHouse-Summary`
-    append_callback([&used_output](const Progress & progress) { used_output.out->onProgress(progress); });
+    append_callback([&used_output](const Progress & progress)
+    {
+        used_output.out->onProgress(progress);
+        auto thread_group = CurrentThread::getGroup();
+        auto peak_memory_usage = thread_group->memory_tracker.getPeak();
+        used_output.out->onMemoryUsage(peak_memory_usage);
```

Comment:
Please clarify to me
- when do you want to collect `peak_memory_usage` for the query
- when do you want to send it to the client
- which clients should receive `peak_memory_usage` and which are not suppose to see it.

The questions arose because I see the inconsistency in the logic, however I might be wrong.
I see that
- 'peak_memory_usage' collected regularly for the query (this is why the data race occurs)
- it is sent only at the end (only at finishSendHeaders)
- only http protocol is affected

Reply:
> when do you want to collect peak_memory_usage for the query

It is already collected (it was implemented) for the query, I just pass it to http. I don't know exactly how it works. The server already displays this value in its log.

> when do you want to send it to the client

I actually need to update once by the end, but there is no such moment here and I send it along with the progress statistics.

> which clients should receive peak_memory_usage and which are not suppose to see it.

I did it for everyone (because I thought it was for information, and no one would be offended). But I don't have exact information on how clients use the X-ClickHouse-Summary field.

> (this is why the data race occurs)

It looks like I forgot the mutex, now I'll try to fix it.


What does it look like
```< Keep-Alive: timeout=10
< X-ClickHouse-Progress: {"read_rows":"3989949","read_bytes":"31919592","written_rows":"0","written_bytes":"0","total_rows_to_read":"20000000","result_rows":"0","result_bytes":"0"}
< X-ClickHouse-Progress: {"read_rows":"8241534","read_bytes":"65932272","written_rows":"0","written_bytes":"0","total_rows_to_read":"20000000","result_rows":"0","result_bytes":"0"}
< X-ClickHouse-Progress: {"read_rows":"12493119","read_bytes":"99944952","written_rows":"0","written_bytes":"0","total_rows_to_read":"20000000","result_rows":"0","result_bytes":"0"}
< X-ClickHouse-Progress: {"read_rows":"16744704","read_bytes":"133957632","written_rows":"0","written_bytes":"0","total_rows_to_read":"20000000","result_rows":"0","result_bytes":"0"}
< X-ClickHouse-Summary: {"read_rows":"20015154","read_bytes":"160121232","written_rows":"0","written_bytes":"0","total_rows_to_read":"20000000","result_rows":"20000000","result_bytes":"160432128","peak_memory_usage":"0"}
```

Reply:
> only http protocol is affected

The task was also about clickhouse-client, and it has a binary protocol. In general, this task consists of two independent tasks: http + clickhouse-client,

Reply:
Yes, I see now, sendProfileEvents works for all other protocols. Good.

Reply:
No I understand why changes in http look so outstanding here. It because CH does't send `ProfileEvents` over http.

Interesting question. It addressed not to you actually. What is the point to send progress as header? If CH is unable to buffer all response then no more headers are sent after data transmission started. Maybe CH really is able to buffer all response in the most cases.

Here I propose to send peak_memory_usage with each `writeHeaderProgress()` call as well. It make sense actually. Because one short peak might not be displayed in memory usage. But it may lead to the OOM.
I think that could make the code more straight.

Reply:
What do you think about  sending peak_memory_usage with each `writeHeaderProgress()` call?

In general I like this PR, and could approve it as it.
But I think it would be even better if just send that info with each progress header, not only the last one.

Reply:
I'm already doing it now (I thought about doing it during development, but then I refused). Today I hope to fix it (writeHeaderProgress + do not display for old servers).

Reply:
Did (send memory in progress)